openrouteservice/src/main/java/org/heigit/ors/matrix/algorithms/core/CoreMatrixAlgorithm.java

```diff
@@ -0,0 +1,538 @@
+                    targetItem.setEdge(msptSubItem.getEdge());
+                    targetItem.setOriginalEdge(msptSubItem.getOriginalEdge());
+                    targetItem.setParent(msptSubItem.getParent());
+                }
+            }
+        }
+    }
+
+    /**
+     * /
+     * /
+     * __________IN-CORE
+     * /
+     * /
+     **/
+
+    /**
+     * Create a Many to Many Dijkstra for the core and downwards phase and run it
+     */
+    private void runPhaseInsideCore() {
+        // Calculate all paths only inside core
+        DijkstraManyToMany algorithm = new DijkstraManyToMany(graph, chGraph, bestWeightMap, bestWeightMapCore, weighting, TraversalMode.NODE_BASED);
+
+        EdgeFilterSequence edgeFilterSequence = new EdgeFilterSequence();
+        edgeFilterSequence.add(this.additionalCoreEdgeFilter);
+
+        algorithm.setEdgeFilter(edgeFilterSequence);
+        algorithm.setTreeEntrySize(this.treeEntrySize);
+        algorithm.setHasTurnWeighting(this.hasTurnWeighting);
+        algorithm.setMaxVisitedNodes(this.maxVisitedNodes);
+        algorithm.setVisitedNodes(this.visitedNodes);
+        algorithm.setTurnWeighting(this.turnWeighting);
+        algorithm.setTargetGraphExplorer(targetGraph.createExplorer());
+        algorithm.setTargetMap(this.targetMap);
+        algorithm.setTargetSet(this.targetSet);
+        algorithm.setSwap(this.swap);
+
+        int[] entryPoints = coreEntryPoints.toArray();
+        int[] exitPoints = coreExitPoints.toArray();
+        algorithm.calcPaths(entryPoints, exitPoints);
+    }
+
+    /**
+     * /
+     * /
+     * __________UTIL
+     * /
+     * /
+     **/
+
+    private boolean isValid(MatrixLocations srcData, MatrixLocations dstData) {
+        return !(!srcData.hasValidNodes() || !dstData.hasValidNodes());
+    }
+
+    /**
+     * Search is more efficient for low source count and high destination count than the other way around.
+     * If there are more sources than destinations, they get swapped and all calculations are done backwards.
+     * The final result gets unswapped to return correct results.
+     * @param srcData original Source data
+     * @param dstData original Destination data
+     * @return
+     */
+    private boolean checkSwapSrcDst(MatrixLocations srcData, MatrixLocations dstData){
+        return(srcData.size() > dstData.size());
+    }
+
+    /**
+     * Invert the results matrix (represented by flattened array) in case src and dst were swapped
+     * @param srcData the original unswapped source data
+     * @param dstData the original unswapped destination data
+     * @param times the swapped array of results
+     * @param distances the swapped array of results
+     * @param weights the swapped array of results
+     * @return array of unswapped result arrays [times, distances, weights]
+     */
+    private float[][] swapResults(MatrixLocations srcData, MatrixLocations dstData, float[] times, float[] distances, float[] weights) {
```

Comment:
These methods could probably be moved as static methods into a separate helper class mentioned [here](https://github.com/GIScience/openrouteservice/pull/959/files#r718595873).

Reply:
The swapping stuff? It is only used here and relatively specific. I am not sure there is any benefit apart from making `CoreMatrixAlgorithm` smaller

Reply:
Yeah, I meant `swapResults`, `isValid`, and `checkSwapSrcDst`. Even though they are specific to `CoreMatrixAlgorithm` they don't depend on the actual object instance. I don't really care, I just thought that moving things around a bit would help in keeping the actual algrithm logic separate from helper functions. It's up to you.

Reply:
The concept of separating out functionality into classes is good. The swapping is not a good candidate for this, though, as it is intertwined with the algorithms in many places. Moving these methods outside of the matrix class leaves both the matrix and the new class with parts of the functionality. This makes little sense, as the new class would be useless by itself, but also the matrix would be useless without the class. It made sense for the turnweighting stuff, but not so much here, I find.

Reply:
Thanks a lot for providing your arguments, they are quite reasonable! Let's leave it then as is :)